fendermint/vm/interpreter/src/fvm/checkpoint.rs

```diff
@@ -212,3 +241,164 @@ where
 
     Ok(PowerTable(power_table))
 }
+
+/// Convert the collaterals and metadata in the membership to the public key and power expected by the system.
+fn membership_to_power_table(m: &Membership, power_scale: PowerScale) -> PowerTable {
+    let mut pt = Vec::new();
+
+    for v in m.validators.iter() {
+        // Ignoring any metadata that isn't a public key.
+        if let Ok(pk) = PublicKey::parse_slice(&v.metadata, None) {
+            let c = u256_to_tokens(v.weight);
+            pt.push(Validator {
+                public_key: ValidatorKey(pk),
+                power: Collateral(c).into_power(power_scale),
+            })
+        }
+    }
+
+    PowerTable(pt)
+}
+
+fn u256_to_tokens(value: U256) -> TokenAmount {
+    let mut bz = [0u8; 32];
+    value.to_big_endian(&mut bz);
+    let atto = BigInt::from_bytes_be(Sign::Plus, &bz);
+    TokenAmount::from_atto(atto)
+}
+
+/// Calculate the difference between the current and the next power table, to return to CometBFT only what changed.
+fn power_diff(current: &PowerTable, next: PowerTable) -> PowerUpdates {
+    let current = current
+        .0
+        .iter()
+        .map(|v| {
+            // Unfortunately the keys don't implement `Hash`.
+            let k = v.public_key.0.serialize();
+            (k, v)
+        })
+        .collect::<HashMap<_, _>>();
+
+    let next = into_power_map(next);
+
+    let mut diff = Vec::new();
+
+    // Validators in current but not in next should be removed.
+    for (k, v) in current.iter() {
+        if !next.contains_key(k) {
+            let delete = Validator {
+                public_key: v.public_key.clone(),
+                power: Power(0),
+            };
+            diff.push(delete);
+        }
+    }
+
+    // Validators in next that differ from current should be updated.
+    for (k, v) in next.into_iter() {
+        let insert = match current.get(&k) {
+            Some(w) if **w == v => None,
+            _ => Some(v),
+        };
+        if let Some(insert) = insert {
+            diff.push(insert);
+        }
+    }
+
+    PowerUpdates(diff)
```

Comment:
Oh, I was having a hard time to see how this diff was applied. So you insert all of the changes beween `current` and `next`, indicating if something wants to be removed by setting their power by `0` (it wasn't obvious to me how you were reflecting validators leaving the membership) :)

Reply:
Yes, that's what CometBFT expects: https://github.com/informalsystems/tendermint-rs/blob/bcc0b377812b8e53a02dff156988569c5b3c81a2/rpc/src/dialect/end_block.rs#L12-L14